### PR TITLE
SOLR-17516: LBHttpSolrClient: support HttpJdkSolrClient (Generic Version)

### PR DESCRIPTION
**Pull request by jdyer1**
Similar to #2805, this is a minimal PR to add Load Balancing functionality for `HttpJdkSolrClient`. Included is test refactoring, adding the *JdkClient as a random choice for the integration test.

Whereas #2805 created a class hierarchy, this makes `LBHttp2SolrClient` generic.  Adding a generic parameter to this class will not break backwards-compatibility.  This seems a better option that #2805.  

One downside is we retain the name `LBHttp2SolrClient` although it is no longer strictly to be used with `Http2SolrClient`.  However, the legacy implementation already uses the preferred name: `LBHttpSolrClient`.  Whenever the legacy implementation is removed, I recommend also renaming `LBHttp2SolrClient` to `LBHttpSolrClient`.

### CODE REVIEW

File: solr/solrj/src/test/org/apache/solr/client/solrj/impl/LBHttp2SolrClientTest.java
Comment (dsmiley) [inline solr/solrj/src/test/org/apache/solr/client/solrj/impl/LBHttp2SolrClientTest.java:54]: perhaps some tests methods shouldn't care in theory what the httpSolrClient impl is exactly but it'd be nice to get one randomly or via parameterization even better.  Not sure if that'd be too difficult.

File: solr/solrj/src/test/org/apache/solr/client/solrj/impl/MockTrustManager.java
Comment (dsmiley) [inline solr/solrj/src/test/org/apache/solr/client/solrj/impl/MockTrustManager.java:34]: nice to see this separated

**Review by dsmiley — COMMENTED** [review]
Overall, I like it!  Better than the subclass one.  I concur with your rename proposals.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpJdkSolrClient.java
Comment (jdyer1) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpJdkSolrClient.java:176]: I found that `LBHttpSolrClient` needs a `SolrServerException` to detect a down node.  Yet we still can throw an `IOException` here during `prepareRequest`.  This roughly mimics the exception handling in `Http2SolrClient`.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java
Comment (jdyer1) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttp2SolrClient.java]: I improved the wording.

File: solr/solrj/src/test/org/apache/solr/client/solrj/impl/LBHttp2SolrClientTest.java
Comment (jdyer1) [reply solr/solrj/src/test/org/apache/solr/client/solrj/impl/LBHttp2SolrClientTest.java:54]: Actually, 4/5 test methods here are using a mock client.  The `withTheseParamNamesInTheUrl` case is being tested with the clients' own unit tests.  I did add a random choice to `LBHttp2SolrClientIntegrationTest` (renamed).  I think this is pretty good coverage already.

File: solr/solrj/src/test/org/apache/solr/client/solrj/impl/MockTrustManager.java
Comment (jdyer1) [reply solr/solrj/src/test/org/apache/solr/client/solrj/impl/MockTrustManager.java:34]: Thank you for taking the time to review.  I'll plan to commit this (main only?; 9.8?) early next week.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpJdkSolrClient.java
Comment (dsmiley) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpJdkSolrClient.java:176]: Okay; just needs a comment because it looks wrong

Comment (dsmiley) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpJdkSolrClient.java:176]: Actually, maybe LBHttpSolrClient should be enhanced to catch the IOException to behave properly here?  Otherwise, it seems wrong just looking at this SolrClient to throw/not-throw based on some wrapping/delegating client unless we spell out these things in the documented semantics of this method.

Comment (jdyer1) [reply solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpJdkSolrClient.java:176]: Indeed, the doc-comment on the `LBSolrClient` request method says,

>  If a request fails due to an IOException, the server is moved to the dead pool

So I restored the previous behavior in `HttpJdkSolrClient` and made the modification in `LBSolrClient`.  This new behavior is covered by `LBHttp2SolrClientIntegrationTest`.

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java
Comment (dsmiley) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBSolrClient.java:704]: It's a shame to see duplication in getRootCause IOException detection with the dedicated clause.  Maybe it'd be cleaner to have another try-catch just around calling request() above that detects the IOException and throws it?  Just a thought; I leave it to you.

**Review by dsmiley — APPROVED** [review]

**Comment by jdyer1** [comment]
@gerlowskija @dsmiley This PR required significant changes to merge with the incoming changes from SOLR-17256.  I would appreciate another round of review before merging.

**Review by dsmiley — APPROVED** [review]
Looks good, though I didn't look too carefully.

**Pull request merged**
